akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowFutureFlowSpec.scala

```diff
@@ -0,0 +1,411 @@
+      fSeq.futureValue should equal(List(0, 1, 2, 3, 4, 99))
+    }
+
+    "propagate upstream failure when future is pre-completed" in assertAllStagesStopped {
+      val (fNotUsed, fSeq) = src10WithFailure()(5)
+        .viaMat {
+          Flow.futureFlow {
+            Future.successful {
+              Flow[Int]
+            }
+          }
+        }(Keep.right)
+        .toMat(Sink.seq)(Keep.both)
+        .run()
+
+      fNotUsed.futureValue should be(NotUsed)
+      fSeq.failed.futureValue should equal(TE("fail on 5"))
+    }
+
+    "propagate upstream failure when future is late-completed" in assertAllStagesStopped {
+      val prFlow = Promise[Flow[Int, Int, NotUsed]]
+      val (fNotUsed, fSeq) = src10WithFailure()(5)
+        .viaMat {
+          Flow.futureFlow(prFlow.future)
+        }(Keep.right)
+        .toMat(Sink.seq)(Keep.both)
+        .run()
+
+      fNotUsed.value should be(empty)
+      fSeq.value should be(empty)
+
+      prFlow.success {
+        Flow[Int]
+      }
+
+      fNotUsed.futureValue should be(NotUsed)
+      fSeq.failed.futureValue should equal(TE("fail on 5"))
+    }
+
+    "handle early upstream error when flow future is pre-completed" in assertAllStagesStopped {
+      val (fNotUsed, fSeq) = Source
+        .failed(TE("not today my friend"))
+        .viaMat {
+          Flow.futureFlow {
+            Future.successful {
+              Flow[Int]
+                .recover {
+                  case TE("not today my friend") => 99
+                }
+                .concat(src10())
+            }
+          }
+        }(Keep.right)
+        .toMat(Sink.seq)(Keep.both)
+        .run()
+
+      fNotUsed.futureValue should be(NotUsed)
+      fSeq.futureValue should equal(99 +: (0 until 10))
+
+    }
+
+    "handle early upstream error when flow future is late-completed" in assertAllStagesStopped {
+      val prFlow = Promise[Flow[Int, Int, NotUsed]]
+      val (fNotUsed, fSeq) = Source
+        .failed(TE("not today my friend"))
+        .viaMat {
+          Flow.futureFlow(prFlow.future)
+        }(Keep.right)
+        .toMat(Sink.seq)(Keep.both)
+        .run()
+
+      fNotUsed.value should be(empty)
+      fSeq.value should be(empty)
+
+      prFlow.success {
+        Flow[Int]
+          .recover {
+            case TE("not today my friend") => 99
+          }
+          .concat(src10())
+      }
+
+      fNotUsed.futureValue should be(NotUsed)
+      fSeq.futureValue should equal(99 +: (0 until 10))
+
+    }
+
+    "handle closed downstream when flow future is pre completed" in assertAllStagesStopped {
+      val (fSeq1, fSeq2) = src10()
+        .viaMat {
+          Flow.futureFlow {
+            Future.successful {
+              Flow[Int].alsoToMat(Sink.seq)(Keep.right)
+            }
+          }
+        }(Keep.right)
+        .mapMaterializedValue(_.flatten)
+        .take(0)
+        .toMat(Sink.seq)(Keep.both)
+        .run()
+
+      fSeq1.futureValue should be(empty)
```

Comment:
Is this true deterministically? The question is if `alsoTo` might already send elements to the second branch of the stream before it receives the cancellation. But maybe it just works out because there are no hidden buffers. In any case, it seems a bit brittle to expect that so there should be a at least a comment that warns about it.

Reply:
let's make sure I understand your concerns, you're talking about the fact there's nop guaranteed order between upstream events and downstream cancellation/failure?
I think it doesn't matter in this case since takeWhile/take will fire a cancellation only after requesting and receiving n messages.

I think your concerns hold for downstream with async boundaries/behaviour or buffering where a cancellation may be fired out of band, but not for this specific test.